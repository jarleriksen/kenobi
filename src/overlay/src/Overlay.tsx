import React, { useContext, useEffect } from 'react';
import { Portal } from '../../portal';
import Transition from 'react-transition-group/Transition';
import styled from 'styled-components';
import { ThemeContext } from '../../theme/src/ThemeConsumer';
import FocusLock from 'react-focus-lock';
import { RemoveScroll } from 'react-remove-scroll';
import OverlayWrapper from './OverlayWrapper';

const ANIMATION_DURATION = 240;

export interface OverlayProps {
  /**
   * Whether or not to show the Overlay.
   */
  isOpen: boolean;

  /**
   * Overlay content
   */
  children: React.ReactNode;

  /**
   * Styles for the OverlayWrapper
   */
  childrenWrapperStyles?: object;

  /**
   * If Overlay should close on click
   */
  shouldCloseOnClick: boolean;

  /**
   * If overlay should close on Escape press
   */
  shouldCloseOnEsc: boolean;

  /**
   * Provided function to set opposite value of isOpen prop
   *
   * This function is used to close the Overlay on Esc or OverlayClick.
   *
   * This is a user provided function.
   */
  onRequestClose: () => void;
}

const Overlay = ({
  isOpen = false,
  shouldCloseOnClick = true,
  shouldCloseOnEsc = true,
  ...props
}: OverlayProps) => {
  const themeContext = useContext(ThemeContext);

  const { children, childrenWrapperStyles, onRequestClose } = props;

  const handleClose = (e) => {
    if (shouldCloseOnEsc && e.key === 'Escape') {
      /**
       * This should be enough to avoid closing
       * nested Overlays. However, it is not.
       * Requires more research, I guess.
       */
      e.stopPropagation();
      onRequestClose();
    }
  };

  const handleOverlayClick = (e) => {
    if (e.target === e.currentTarget && shouldCloseOnClick) {
      onRequestClose();
    }
  };

  return (
    <Portal>
      <Transition appear unmountOnExit timeout={ANIMATION_DURATION} in={isOpen}>
        {(state) => (
          <FocusLock returnFocus>
            <RemoveScroll>
              <OverlayWrapper
                theme={themeContext}
                onClick={handleOverlayClick}
                onKeyDown={handleClose}
                style={{ ...childrenWrapperStyles }}
                state={state}
                data-testid="kenobi-overlay">
                {typeof children === 'function'
                  ? children({ state })
                  : children}
              </OverlayWrapper>
            </RemoveScroll>
          </FocusLock>
        )}
      </Transition>
    </Portal>
  );
};

export default Overlay;
