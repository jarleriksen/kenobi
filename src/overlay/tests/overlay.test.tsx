import React from 'react';
import { render } from '@testing-library/react';
import Overlay from '../src/Overlay';
import '@testing-library/jest-dom/extend-expect';

describe('Overlay', () => {
  test('is in portal when open', () => {
    const { getByTestId } = render(
      <Overlay
        isOpen={true}
        childrenWrapperStyles={{}}
        onRequestClose={() => {}}
        shouldCloseOnEsc={true}
        shouldCloseOnClick={true}>
        <div />
      </Overlay>,
    );

    expect(getByTestId('kenobi-overlay')).toBeInTheDocument();
  });

  test('is not in portal when closed', () => {
    const { getByTestId } = render(
      <Overlay
        isOpen={false}
        childrenWrapperStyles={{}}
        onRequestClose={() => {}}
        shouldCloseOnEsc={true}
        shouldCloseOnClick={true}>
        <div />
      </Overlay>,
    );

    const portal = getByTestId('kenobi-portal');
    const overlay: HTMLElement | null = document.querySelector(
      '[data-testid="kenobi-overlay"]',
    );

    expect(portal).not.toContainElement(overlay);
  });
});
