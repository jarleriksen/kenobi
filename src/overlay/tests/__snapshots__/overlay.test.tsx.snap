// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay is in portal when open 1`] = `<DocumentFragment />`;

exports[`Overlay is not in portal when closed 1`] = `<DocumentFragment />`;
